### ISSUE TITLE
Question:  How do I configure TexStudio to run splitindex?

### ISSUE THREAD

**Issue by mkstreet** [texstudio-org/texstudio#1073]
### Environment
- TeXstudio: 2.12.6 (hg 6631:16db084dae24)
- Qt: Version 5.9.5, compiled with Qt 5.9.3 R
- OS: Ubuntu 18.04 
- TeX distribution: texlive 2019

### Expected behavior

How do I configure the TexStudio options so that splitindex will be run?

**Comment by MeanSquaredError** [comment]
I am not familiar with the splitindex package, so I can only provide generic advice:

It seems that splitindex.pl is a drop-in replacement for `makeindex`, so you have to replace `makeindex` with `splitindex.pl` in `Options / Configure TeXstudio... / Commands / Makeindex`

Hope that helps.

**Comment by mkstreet** [comment]
> I am not familiar with the splitindex package, so I can only provide generic advice:
> 
> It seems that splitindex.pl is a drop-in replacement for `makeindex`, so you have to replace `makeindex` with `splitindex.pl` in `Options / Configure TeXstudio... / Commands / Makeindex`
> 
> Hope that helps.

I need both makeindex and splitindex.

Right now, manually, I run splitindex from the command line with  myfilename.idx as its input.
Then I have to build the pdf again to pull in the splitindex output.

 Ideally, I would just like TexStudio to run splitindex after makeindex has run.
 It runs makeindex now only.

**Comment by aubreyz** [comment]
I think you also need shell-escape so compile with something like:

     latexmk -xelatex -synctex=1 --shell-escape -interaction=nonstopmode %

**Comment by MeanSquaredError** [comment]
@mkstreet 
Are you sure that you really need both `makeindex` (the executable file) and `splitindex` ?
The latex compiler (e.g. `pdflatex`) will create the `.idx` file even if you don't run `makeindex`. 

Or do you mean that you need the final index file both as a single `.ind` file and also split into multiple `.ind` files?

**Comment by mkstreet** [comment]
I'm sorry, I don't know about how the .idx file is normally created, I guess.

The splitindex functionality relies on the .idx file being created prior.  Then when I run splitindex at the command line, it takes that normal, singular .idx file as input and creates (splits) separate .idx files, one for each of the indexes defined in the .tex document.  As far as I know, only those separated .idx files are used, pulled-in to, the pdf creation.

You mentioned the compiler, so it may be important to know that I am using:  
    % !TeX program = xelatex

**Comment by MeanSquaredError** [comment]
The .idx file(s) will be created by whichever TeX compiler you use (in your case `xelatex`) as long as you use the `makeindex` or `splitindex` macros in your .tex source file. You don't have to run the `makeindex` binary to get the .idx file(s). So you can just configure TeXstudio to run `splitindex.pl` (or whichever binary the `splitindex` package comes with) and it will build the `.ind` files for you.

**Comment by MeanSquaredError** [comment]
Now, after re-reading your message, I am not sure that I fully understand your setup. After running splitindex from the command line, do you run `makeindex` from the command line too?

**Comment by mkstreet** [comment]
The .idx file is being created out of the normal compile process.  However that works.

Then I am running the splitindex from the command line.

I am not running the makeindex - I assumed (sorry) that was being run automatically by the compile process.

So, what I want is the running of splitindex to become part of the normal compile stream so that I do not have to remember to go to the command line and run it separately myself.

**Comment by MeanSquaredError** [comment]
You can configure `splitindex` as a replacement for `makeindex` in 
`Options / Configure TeXstudio... / Commands / Makeindex`

Then add the makeindex command to the `Default compiler` in 
`Options / Configure TeXstudio... / Build / Default Compiler`
E.g. if your default compiler was:
`txs:///xelatex`
You can set it to
`txs:///xelatex | txs:///makeindex`
That should take care of running txs:///makeindex (actually splitindex) automatically after running xelatex

**Comment by mkstreet** [comment]
My default compiler is  txs:///pdflatex

xelatex is being activated as an exception for this document using the magic comments at the start of the document:
   % !TeX program = xelatex

Is there another place in the build options I can change?  I see there is a USER COMMAND, maybe I can define that somehow?

**Comment by aubreyz** [comment]
What's wrong with shell escape?

**Comment by mkstreet** [comment]
> What's wrong with shell escape?

I don't know how to do that. How is it done?

**Comment by MeanSquaredError** [comment]
When you are using a magic comment to specify the compiler, e.g.
`% !TeX program = xelatex`
TXS will completely ignore the the command(s) specified in `Default Compiler` and will replace it temporarily with the program that you specified, in this case `txs:///xelatex`

If you insist on using magic comments to enforce a different compiler (in this case `xelatex`, then I can suggest creating a user command, e.g.:
`xelatex_and_index:` and then set it to something like `txs:///xelatex | txs:///makeindex`
Then use a magic comment like:
`% !TeX program = xelatex_and_index`

Alternatively you could create shell scripts that would run the corresponding .tex compiler and then would run `splitindex`. You would have one such script per compiler (`pdflatex`, `xelatex`, etc.), then you would specify these scripts as compilers in the commands section.

However the first approach is probably cleaner.

**Comment by mkstreet** [comment]
Under   options->configure->build->user commands

I added the xelatex_and_index and defined it as  txs:///xelatex | txs:///makeindex

I changed the magic comments and tried to compile.

It tells me
 Error: could not start the command xelatex_and_index

I get this same error if I put in the wrong command name in the magic comments.
I get this same error if I change its definition to be only    txs:///xelatex

**Comment by MeanSquaredError** [comment]
From the information that you provided it seems that you are using a relatively old version of TXS (2.12.6). If I remember correctly versions before 2.12.16 had problems with user commands.

Can you try a newer version, preferably one of the latest 3.0.0 (available for download from www.texstudio.org)? If it still does not work please upload your texstudio.ini and a sample .tex file that uses a magic comment and I will take a look on my computer.

**Comment by mkstreet** [comment]
Ubuntu hasn't included a later version than 2.12.6 in their releases yet.
But I got 2.12.22 from texstudio.org and updated my system to that.  

That's the latest release on the website.  Later releases would have to come direct from git hub and are marked as alpha.

I zipped my files and attached them.
[ezyzip.zip](https://github.com/texstudio-org/texstudio/files/4581874/ezyzip.zip)

When I looked inside the texstudio.ini, I don't see a definition of my user command list, but in texstudio under configure options -- it is there.  I sent screen shots in the zip.

**Comment by MeanSquaredError** [comment]
OK, I checked the TXS code and it seems that TXS insists on user commands having their IDs starting with `user`. Having that in mind you need to configure TXS as following:

1. Add a custom user command named `user0:xelatexsplit` and with definition 
`txs:///xelatex | txs:///makeindex`
Basically the part in the name after the `:` is pretty much irrelevant, but using `xelatexsplit` is a good reminder of the command purpose.
2. Set the following magic comment in your .tex file:
`% !TeX TS-program = txs:///user0`

I tested this with the latest TXS from the github repository (effectively TXS 3.0.0-alpha) and it worked fine. However I think it should work with TXS 2.12.22 too. I do not recommend using 2.12.6 since IIRC it has some problems with user commands.

**Comment by MeanSquaredError** [comment]
Actually it just occurred to me that you can specify the whole command chain as the magic comment and skip the user command altogether. So just specifying:
`% !TeX TS-program = txs:///xelatex | txs:///makeindex | txs:///view-pdf-external`
at the top of your .tex file should do the trick and you don't need a user command.

And for the other .tex files that do not need a special compiler just set 
`Options / Configure TeXstudio... / Build / Default Compiler`
to
`txs:///pdflatex | txs:///makeindex`

**Comment by mkstreet** [comment]
It works!

Using your information, thank you, I wound up setting it up a little differently.

1. I restored   Configure->Commands->makeindex   back to the system default, so that it is not adulterated anymore

2. I defined a user command as:    `user0:mksplitindex `   with a definition of  ` splitindex %.idx`

3. I changed this document's magic comments to:
    `  % !TeX TS-program = txs:///xelatex |txs:///user0 | txs:///view`

I am running TXS 2.12.22 now, and I will stay with that.  When vers 3.xx.xx moves out of alpha and to the texstudio.org website, I will upgrade.

Since the TexStudio settings are now becoming more than just cosmetic usage of TexStudio, I think I want to have the settings included in my project version control.  Is saving off the texstudio.ini file enough - or do I need to consider other configuration files too?
